src/hardware/mixer.cpp

```diff
@@ -789,37 +718,23 @@ static void SDLCALL MIXER_CallBack(MAYBE_UNUSED void *userdata, Uint8 *stream, i
 	mixer.needed -= reduce;
 	pos = mixer.pos;
 	mixer.pos = (mixer.pos + reduce) & MIXER_BUFMASK;
-	if(need != reduce) {
-		while (need--) {
-			Bitu i = (pos + (index >> INDEX_SHIFT_LOCAL )) & MIXER_BUFMASK;
-			index += index_add;
-			sample=mixer.work[i][0]>>MIXER_VOLSHIFT;
-			*output++=MIXER_CLIP(sample);
-			sample=mixer.work[i][1]>>MIXER_VOLSHIFT;
-			*output++=MIXER_CLIP(sample);
-		}
-		/* Clean the used buffer */
-		while (reduce--) {
-			pos &= MIXER_BUFMASK;
-			mixer.work[pos][0]=0;
-			mixer.work[pos][1]=0;
-			pos++;
-		}
-	} else {
-		while (reduce--) {
-			pos &= MIXER_BUFMASK;
-			sample=mixer.work[pos][0]>>MIXER_VOLSHIFT;
-			*output++=MIXER_CLIP(sample);
-			sample=mixer.work[pos][1]>>MIXER_VOLSHIFT;
-			*output++=MIXER_CLIP(sample);
-			mixer.work[pos][0]=0;
-			mixer.work[pos][1]=0;
-			pos++;
-		}
+
+	while (reduce--) {
+		pos &= MIXER_BUFMASK;
+		sample = mixer.work[pos][0] >> MIXER_VOLSHIFT;
+		queue_buffer[buf_idx++] = MIXER_CLIP(sample);
+		sample = mixer.work[pos][1] >> MIXER_VOLSHIFT;
+		queue_buffer[buf_idx++] = MIXER_CLIP(sample);
+		mixer.work[pos][0] = 0;
+		mixer.work[pos][1] = 0;
+		pos++;
 	}
-}
 
-#undef INDEX_SHIFT_LOCAL
+	const uint32_t size = len * MIXER_SSIZE;
+	const auto res = SDL_QueueAudio(mixer.sdldevice, queue_buffer.data(), size);
```

Comment:
We've got MIXER_SSIZE = 4 and `* 2` in the assert above `assert(static_cast<std::size_t>(len) * 2 <= queue_buffer.size());`

I wonder if these two can be collapsed into `constexpr sample_bytes = sizeof(sample)`, and the assert moved just prior to the `SDL_QueueAudio`.  Would be nice to eliminate the hardcoded 4 and 2's.

Reply:
Maybe `sample` can be eliminated as a middleman too? 
I think this is readable as-is. 
```c++
	queue_buffer[buf_idx++] = MIXER_CLIP(mixer.work[pos][0] >> MIXER_VOLSHIFT);
	queue_buffer[buf_idx++] = MIXER_CLIP(mixer.work[pos][1] >> MIXER_VOLSHIFT);
```

Reply:
The kicker is, the sample size is 4, since SDL treats 16 bits * 2 channels = 1 sample of 4 bytes.

So, my logic was, the max samples we support give 100ms latency at 48kHz would be a max 8192, giving 32k total bytes, and the existing constant `MIXER_BUFSIZE` is 16k, so that's where the * 2 came from.

Just typing this out, yeah, we need to simplify. I propose a MAX_QUEUE_SAMPLES = 8192 and then just leave the MIXER_SSIZE as-is.